meshmode/discretization/connection/modal.py

```diff
@@ -0,0 +1,373 @@
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+import numpy.linalg as la
+import modepy as mp
+
+from meshmode.array_context import make_loopy_program
+from meshmode.dof_array import DOFArray
+from meshmode.discretization import (
+    NodalElementGroupBase,
+    ModalElementGroupBase,
+    InterpolatoryElementGroupBase)
+from meshmode.discretization.poly_element import QuadratureSimplexElementGroup
+from meshmode.discretization.connection.direct import DiscretizationConnection
+
+from pytools import memoize_in, keyed_memoize_in
+from pytools.obj_array import obj_array_vectorized_n_args
+
+
+class NodalToModalDiscretizationConnection(DiscretizationConnection):
+    """A concrete subclass of :class:`DiscretizationConnection`, which
+    maps nodal data to its modal representation. This connection can
+    be used with both unisolvent and non-unisolvent element groups.
+
+    .. note::
+
+        This connection requires that both nodal and modal discretizations
+        are defined on the *same* mesh.
+
+    .. attribute:: from_discr
+
+        an instance of :class:`meshmode.discretization.Discretization` containing
+        :class:`meshmode.discretization.NodalElementGroupBase` element groups
```

Comment:
```suggestion
        :class:`~meshmode.discretization.NodalElementGroupBase` element groups
```
? These names can get a bit long when looking through the generated docs.